kotlin/workflow-ui-core/src/main/java/com/squareup/workflow/ui/BackStackScreen.kt

```diff
@@ -16,17 +16,80 @@
 package com.squareup.workflow.ui
 
 /**
- * @param stack: screens that are being or have been displayed, ending in the current screen.
- * Guaranteed not to be empty.
+ * Represents an active screen ([top]), and a set of previously visited
+ * screens to which we may return ([backStack]). By rendering the entire
+ * history we allow the UI to do things like maintain cached view state,
+ * implement drag-back gestures without waiting for the workflow, etc.
+ *
+ * Effectively a list that can never be empty.
  */
-data class BackStackScreen<StackedT : Any>(val stack: List<StackedT>) {
-  constructor(only: StackedT) : this(listOf(only))
-  constructor(vararg stack: StackedT) : this(stack.toList())
+class BackStackScreen<StackedT : Any>(
+  bottom: StackedT,
+  rest: List<StackedT>
+) {
+  /**
+   * Creates a screen with elements listed from the [bottom] to the top.
+   */
+  constructor(
+    bottom: StackedT,
+    vararg rest: StackedT
+  ) : this(bottom, rest.toList())
+
+  /**
+   * Creates a screen with the frames of the given [backStack], capped with [top].
+   * [backStack] may be empty.
+   */
+  constructor(
+    backStack: List<StackedT>,
+    top: StackedT
+  ) : this(
+      bottom = backStack.firstOrNull() ?: top,
+      rest = backStack.takeIf { it.isNotEmpty() }
+          ?.let { it.subList(1, it.size) + top }
+          ?: emptyList()
+  )
+
+  val frames: List<StackedT> = listOf(bottom) + rest
```

Comment:
These are doing a bunch of unnecessary list copies, might be better to use a mutable list to massage the lists.

Reply:
We talked, don't actually see any real improvements to be made.